From af39a960a424d7532733a72d207ee1f084b81e96 Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Wed, 17 Feb 2010 20:18:10 +0100
Subject: [PATCH] mount-root: do not pollute init arguments

and do not remount if no new rootopts set in /etc/fstab
---
 modules.d/95rootfs-block/mount-root.sh | 44 ++++++++++++++++++----------------
 1 file changed, 23 insertions(+), 21 deletions(-)

diff --git a/modules.d/95rootfs-block/mount-root.sh b/modules.d/95rootfs-block/mount-root.sh
index 62290ab..3c18406 100755
--- a/modules.d/95rootfs-block/mount-root.sh
+++ b/modules.d/95rootfs-block/mount-root.sh
@@ -2,9 +2,29 @@
 
 . /lib/dracut-lib.sh
 
+filter_rootopts() {
+    rootopts=$1
+    # strip ro and rw options
+    local OLDIFS=$IFS
+    IFS=,
+    set -- $rootopts
+    IFS=$OLDIFS
+    local v
+    while [ $# -gt 0 ]; do
+        case $1 in
+            rw|ro);;
+            *)
+                v="$v,${1}";;
+        esac
+        shift
+    done
+    rootopts=${v#,}
+    echo $rootopts
+}
+
 if [ -n "$root" -a -z "${root%%block:*}" ]; then
     mount -t ${fstype:-auto} -o "$rflags" "${root#block:}" "$NEWROOT" \
-        && ROOTFS_MOUNTED=yes
+        && ROOTFS_MOUNTED=yes 
 
     if ! getarg rd_NO_FSTAB \
       && ! getarg rootflags \
@@ -23,27 +43,9 @@ if [ -n "$root" -a -z "${root%%block:*}" ]; then
             fi
 	done < "$NEWROOT/etc/fstab"
 
+	rootopts=$(filter_rootopts $rootopts)
 
-	# strip ro and rw options
-	{
-	    local OLDIFS=$IFS
-	    IFS=,	
-	    set -- $rootopts
-	    IFS=$OLDIFS
-	    local v
-	    while [ $# -gt 0 ]; do
-		case $1 in 
-		    rw|ro);;
-		    *)
-			v="$v,${1}";;
-		esac
-		shift
-	    done
-	    rootopts=${v#,}
-	}
-
-
-	if [ "$rootopts" != "defaults" ]; then
+	if [ -n "$rootopts" -a "$rootopts" != "defaults" ]; then
             umount $NEWROOT
             info "Remounting ${root#block:} with -o $rootopts,$rflags"
             mount -t "$rootfs" -o "$rflags","$rootopts" \
